### PR TITLE
feat(KS+): More modeled errors

### PR DESCRIPTION
**Pull request by texastony**
Give ourselves room to:
1. Improve the KMS errors during a KMS Key Transition or Mutation in general
2. Resume Mutations and Describe Mutation need some Storage Operations
3. The Storage Operations need to discern why a write failed.

More on (3.):
Write Operations can fail because something else 
modified or deleted an item since it was read,
OR 
Mutation writes can fail because 
the M-Lock and M-Token disagree.

The two conditions have different recovery methods.
i.e: If the Lock/Token is failing the request, 
do not retry with this Token, 
get a fresh one, 
or "criticize" the Lock.

Thus, we need to differentiate between them.

Since we cannot add or change operations on the Storage Model
after it has been released,
we MUST get these Storage changes in.

More on (1.):
KMS Access Denied errors are, 
unfortunately, not modeled in their Smithy Model,
and therefore represented in Dafny as Opaque Errors.

We will need to either benerate or 
implement additional logic in Smithy-Dafny to
fully handle Mutation errors from KMS.
(See https://github.com/smithy-lang/smithy-dafny/issues/609)

But for the immediate future,
we can consider returning the error shapes
introduced here when an opaque error is encountered.

### CODE REVIEW

**Comment by texastony** [comment]
Huh... all my efforts to have a Git History that is consistent have been for naught...
I wonder what mistake I made...

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/Model/Storage.smithy [hunk 1]
```diff
@@ -470,16 +495,76 @@ structure WriteMutatedVersionsInput {
...
+Otherwise, returns nothing.")
+operation GetMutationLock {
+  input: GetMutationLockInput
+  output: GetMutationLockOutput
+  errors: [KeyStorageException]
+}
+structure GetMutationLockInput {
+  @documentation("The Branch Key to check for a Mutation Lock.")
+  @required
+  Identifier: String
```
[the remark this answers is not in the document]
Comment (texastony) [reply AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/Model/Storage.smithy:516]: Resolved in 1591f3fa2

**Comment by texastony** [comment]
@josecorella has reviewed everything but the last 7 commits.
@josecorella  you can view 
those commits via this link:
https://github.com/aws/aws-cryptographic-material-providers-library/pull/754/files/994813f64dc7ddee1b577d2f16b5437372ed6ff6..0ebdc5d52bc4fb3460d74712df352ec5a22b0716

File: AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/Model/AwsCryptographyKeyStoreTypes.dfy [hunk 2]
```diff
@@ -39,7 +39,7 @@ module {:extern "software.amazon.cryptography.keystore.internaldafny.types" } Aw
     nameonly branchKey: Secret
   )
   datatype ClobberMutationLockInput = | ClobberMutationLockInput (
```
Comment (josecorella) [inline AwsCryptographicMaterialProviders/dafny/AwsCryptographyKeyStore/Model/AwsCryptographyKeyStoreTypes.dfy:41]: just making sure, for the smithy models, we are opting to go against smithy style convention, right?
https://smithy.io/2.0/guides/style-guide.html#naming

**Pull request merged**